aspnetcore/fundamentals/app-state.md

```diff
@@ -1,188 +1,312 @@
 
-The server uses the `IdleTimeout` property to determine how long a session can be idle before its contents are abandoned. This property is independent of the cookie expiration. Each request that passes through the Session middleware (read from or written to) resets the timeout.
+::: moniker range=">= aspnetcore-2.0"
 
-Because `Session` is *non-locking*, if two requests both attempt to modify the contents of session, the last one overrides the first. `Session` is implemented as a *coherent session*, which means that all the contents are stored together. Two requests that are modifying different parts of the session (different keys) might still impact each other.
+[!code-csharp[](app-state/samples/2.x/SessionSample/Extensions/SessionExtensions.cs?name=snippet1)]
 
-### Set and get Session values
+::: moniker-end
+
+::: moniker range="< aspnetcore-2.0"
+
+[!code-csharp[](app-state/samples/1.x/SessionSample/Extensions/SessionExtensions.cs?name=snippet1)]
+
+::: moniker-end
+
+The following example shows how to set and get a serializable object with the extension methods:
 
-Session is accessed from a Razor Page or view with `Context.Session`:
+::: moniker range=">= aspnetcore-2.0"
 
-[!code-cshtml[](app-state/sample/src/WebAppSessionDotNetCore2.0App/Views/Home/About.cshtml)]
+[!code-csharp[](app-state/samples/2.x/SessionSample/Pages/Index.cshtml.cs?name=snippet2)]
 
-Session is accessed from a `PageModel` class or controller with `HttpContext.Session`. This property is an [ISession](/dotnet/api/microsoft.aspnetcore.http.isession) implementation.
+::: moniker-end
 
-The following example shows setting and getting an int and a string:
+::: moniker range="< aspnetcore-2.0"
+
+[!code-csharp[](app-state/samples/1.x/SessionSample/Controllers/HomeController.cs?name=snippet2&highlight=4,12)]
+
+::: moniker-end
+
+## TempData
+
+ASP.NET Core exposes the [TempData property of a Razor Pages page model](/dotnet/api/microsoft.aspnetcore.mvc.razorpages.pagemodel.tempdata) or [TempData of an MVC controller](/dotnet/api/microsoft.aspnetcore.mvc.controller.tempdata). This property stores data until it's read. The [Keep](/dotnet/api/microsoft.aspnetcore.mvc.viewfeatures.itempdatadictionary.keep) and [Peek](/dotnet/api/microsoft.aspnetcore.mvc.viewfeatures.itempdatadictionary.peek) methods can be used to examine the data without deletion. TempData is particularly useful for redirection when data is required for more than a single request. TempData is implemented by TempData providers using either cookies or session state.
+
+### TempData providers
 
-[!code-csharp[](app-state/sample/src/WebAppSession/Controllers/HomeController.cs?range=8-27,49)]
+::: moniker range=">= aspnetcore-2.0"
 
-If you add the following extension methods, you can set and get serializable objects to Session:
+In ASP.NET Core 2.0 or later, the cookie-based TempData provider is used by default to store TempData in cookies.
```

Comment:
Should TempData be code-fenced all throughout?

Reply:
Outside of the property explicitly named "TempData," it's also the attribute and part of the backing dictionary name. It seems to encompass several parts of the API, thus it seems to reflect a named component of the framework. Therefore, I was leaning toward just leaving it  as text. I think it reads a little better this way, and things will get a little *blocky* in this section if we code-fence all of these. @Rick-Anderson @rachelappel Thoughts?